libs/vkd3d/state.c

```diff
@@ -3625,6 +3676,57 @@ static uint32_t vkd3d_bindless_state_get_bindless_flags(struct d3d12_device *dev
     return flags;
 }
 
+static bool vkd3d_bindless_supports_mutable_type(struct d3d12_device *device)
+{
+    VkDescriptorType descriptor_types[VKD3D_MAX_MUTABLE_DESCRIPTOR_TYPES];
+    const struct vkd3d_vk_device_procs *vk_procs = &device->vk_procs;
+    VkDescriptorSetLayoutBindingFlagsCreateInfoEXT binding_flags;
+    VkMutableDescriptorTypeCreateInfoVALVE mutable_info;
+    VkDescriptorSetLayoutCreateInfo set_layout_info;
+    VkMutableDescriptorTypeListVALVE mutable_list;
+    VkDescriptorSetLayoutSupport supported;
+    VkDescriptorSetLayoutBinding binding;
+
+    const VkDescriptorBindingFlagsEXT binding_flag =
+            VK_DESCRIPTOR_BINDING_UPDATE_AFTER_BIND_BIT_EXT |
+            VK_DESCRIPTOR_BINDING_VARIABLE_DESCRIPTOR_COUNT_BIT_EXT |
+            VK_DESCRIPTOR_BINDING_PARTIALLY_BOUND_BIT_EXT |
+            VK_DESCRIPTOR_BINDING_UPDATE_UNUSED_WHILE_PENDING_BIT_EXT;
+
+    if (!device->device_info.mutable_descriptor_features.mutableDescriptorType)
+        return false;
+
+    mutable_info.sType = VK_STRUCTURE_TYPE_MUTABLE_DESCRIPTOR_TYPE_CREATE_INFO_VALVE;
+    mutable_info.pNext = NULL;
+    mutable_info.pMutableDescriptorTypeLists = &mutable_list;
+    mutable_info.mutableDescriptorTypeListCount = 1;
+
+    mutable_list.descriptorTypeCount = vkd3d_bindless_build_mutable_type_list(device, descriptor_types);
+    mutable_list.pDescriptorTypes = descriptor_types;
+
+    binding.binding = 0;
+    binding.descriptorType = VK_DESCRIPTOR_TYPE_MUTABLE_VALVE;
+    binding.descriptorCount = d3d12_max_descriptor_count_from_heap_type(D3D12_DESCRIPTOR_HEAP_TYPE_CBV_SRV_UAV);
+    binding.pImmutableSamplers = NULL;
+    binding.stageFlags = VK_SHADER_STAGE_ALL;
+
+    binding_flags.sType = VK_STRUCTURE_TYPE_DESCRIPTOR_SET_LAYOUT_BINDING_FLAGS_CREATE_INFO_EXT;
+    binding_flags.pNext = &mutable_info;
+    binding_flags.bindingCount = 1;
+    binding_flags.pBindingFlags = &binding_flag;
+
+    set_layout_info.sType = VK_STRUCTURE_TYPE_DESCRIPTOR_SET_LAYOUT_CREATE_INFO;
+    set_layout_info.pNext = &binding_flags;
+    set_layout_info.flags = VK_DESCRIPTOR_SET_LAYOUT_CREATE_UPDATE_AFTER_BIND_POOL_BIT;
+    set_layout_info.bindingCount = 1;
+    set_layout_info.pBindings = &binding;
+
+    supported.sType = VK_STRUCTURE_TYPE_DESCRIPTOR_SET_LAYOUT_SUPPORT;
+    supported.pNext = NULL;
+    VK_CALL(vkGetDescriptorSetLayoutSupport(device->vk_device, &set_layout_info, &supported));
+    return supported.supported == VK_TRUE;
```

Comment:
do we need to check for the host local one as well or is that guaranteed to be supported?

Reply:
Guaranteed by spec, but I'll check that as well for completeness' sake.

Reply:
Fixed.